README.md
# greekocr

CHANGES
Changelog of the GreekOCR Toolkit for Gamera
============================================

Version 2.0.0, Mai 03, 2022
---------------------------

 - Ported to Python 3 and Gamera 4


Version 1.0.1, Sep 19, 2011
---------------------------

 - FSF address corrected


Version 1.0.0, Mar 09, 2011
---------------------------

 - first creation


scripts/greekocr4gamera.py
#!/usr/bin/env python
# -*- mode: python; indent-tabs-mode: nil; tab-width: 3 -*-
# vim: set tabstop=3 shiftwidth=3 expandtab:

# Copyright (C) 2010-2011 Christian Brandt
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


# This just simply runs the greekocr toolkits main function
import sys

def usage():
   usage = "Usage:\n"
   usage += "  greekocr4gamera.py -x <traindata> [options] <imagefile>\n"
   usage += "\n"
   usage += "  Options:\n"
   usage += "   --wholistic          wholistic segmentation mode (default)\n"
   usage += "   -w                   short for --wholistic\n"
   usage += "   --separatistic       separatistic segmentation mode\n"
   usage += "   -s                   short for --separatistic\n"
   usage += "\n"
   usage += "   --unicode <file>     specify filename for unicode output\n"
   usage += "   -u <file>            short for --unicode\n"
   usage += "   --teubner <file.tex> specify filename for teubner TeX output\n"
   usage += "   -t <file>            short for --teubner\n"
   usage += "\n"
   usage += "   --deskew             do a skew correction (recommended)\n"
   usage += "   --filter             filter out very large (images) and very\n" 
   usage += "                        small components (noise)\n"
   usage += "\n"
   usage += "   --debug              save debug-images\n"
   usage += "                        debug_lines.png debug_words.png debug_chars.png\n"
   usage += "   -d                   short for --debug\n"
   sys.stderr.write(usage)




options = {}
args = sys.argv[1:]

i = 0
while i < len(args):
   if args[i] in ("-x", "--trainingdata"):
      i += 1
      options["trainingdata"] = args[i]
   elif args[i] in ("--help", "-h"):
      usage()
   elif args[i] in ("--wholistic", "-w"):
      options["mode"] = "wholistic"
   elif args[i] in ("--separatistic", "-s"):
      options["mode"] = "separatistic"
   elif args[i] in ("-u","--unicode"):
      i += 1
      options["unicodeoutfile"] = args[i]
   elif args[i] in ("-t", "--teubner"):
      i += 1
      options["teubneroutfile"] = args[i]
   elif args[i] in ("-d", "--debug"):
      options["debug"] = True
   elif args[i] in ("--deskew"):
      options["deskew"] = True
   elif args[i] in ("--filter"):
      options["filter"] = True
   else:
      options["imagefile"] = args[i]
   i += 1

if "trainingdata" not in options:
   print("No Trainingdata given")
   usage()
   exit(1)
if "mode" not in options:
   options["mode"] = "wholistic"
if "imagefile" not in options:
   print("No filename given")
   usage()
   exit(2)


from gamera.core import *
from gamera.plugins.listutilities import median
from gamera.toolkits.greekocr import GreekOCR

g = GreekOCR()
g.mode = options["mode"]
g.load_trainingdata(options["trainingdata"])

image = load_image(options["imagefile"])
if image.data.pixel_type != ONEBIT:
   image = image.to_onebit()

if "filter" in options and options["filter"] == True:
    count = 0
    ccs = image.cc_analysis()
    if "debug" in options and options["debug"] == True:
       print(("filter started on",len(ccs) ,"elements..."))
    median_black_area = median([cc.black_area()[0] for cc in ccs])
    for cc in ccs:
      if(cc.black_area()[0] > (median_black_area * 10)):
        cc.fill_white()
        del cc
        count = count + 1
    for cc in ccs:
      if(cc.black_area()[0] < (median_black_area / 10)):
        cc.fill_white()
        del cc
        count = count + 1
    if "debug" in options and options["debug"] == True:
       print(("filter done.",len(ccs)-count,"elements left."))


if "deskew" in options and options["deskew"] == True:
  #from gamera.toolkits.otr.otr_staff import *
  if "debug" in options and options["debug"] == True:
    print("\ntry to skew correct...")
  rotation = image.rotation_angle_projections(-10,10)[0]
  img = image.rotate(rotation,0)
  if "debug" in options and options["debug"] == True:
    print(("rotated with",rotation,"angle"))



output = g.process_image(image)
if "debug" in options and options["debug"] == True:
   g.save_debug_images()

if "unicodeoutfile" in options:
   g.save_text_unicode(options["unicodeoutfile"])
elif "teubneroutfile" in options:
   g.save_text_teubner(options["teubneroutfile"])
else:
   print(output)



gamera/toolkits/greekocr/compare.py
# -*- mode: python; indent-tabs-mode: nil; tab-width: 3 -*-
# vim: set tabstop=3 shiftwidth=3 expandtab:

# Copyright (C) 2010-2011 Christian Brandt
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import unicodedata
import codecs

def levenshtein(s1, s2):
   """Computes the Levenshtein distance (aka edit distance) between
the two Unicode strings *s1* and *s2*. Signature:

  ``levenshtein(s1, s2)``

This implementation differs from the plugin *edit_distance* in the Gamera
core in two points:

 - the Gamera core function is implemented in C++ and currently only
   works with ASCII strings

 - this implementation is written in pure Python and therefore somewhat
   slower, but it works with Unicode strings.

For details about the algorithm see
http://en.wikibooks.org/wiki/Algorithm_implementation/Strings/Levenshtein_distance
"""
   if len(s1) < len(s2):
      return levenshtein(s2, s1)
   if not s1:
      return len(s2)
 
   previous_row = list(range(len(s2) + 1))
   for i, c1 in enumerate(s1):
      current_row = [i + 1]
      for j, c2 in enumerate(s2):
         insertions = previous_row[j + 1] + 1 # j+1 instead of j since previous_row and current_row are one character longer
         deletions = current_row[j] + 1       # than s2
         substitutions = previous_row[j] + (c1 != c2)
         current_row.append(min(insertions, deletions, substitutions))
      previous_row = current_row
   return previous_row[-1]
 


def levenshtein_multi_unicode(str1, str2):
   #remove linebreaks
   str1 = str1.replace("\n", "")
   str2 = str2.replace("\n", "")
   
   #remove spaces
   str1 = str1.replace(" ", "")
   str2 = str2.replace(" ", "")

   str1_n = unicodedata.normalize("NFD", str1)
   str2_n = unicodedata.normalize("NFD", str2)
   
   return levenshtein(str1_n, str2_n), len(str1_n), len(str2_n)



def errorrate(groundtruth, ocr):
   """For the two given Unicode strings, the edit distance divided by the
length of the first string is returned.
Signature:

  ``errorrate(groundtruth, ocr)``

"""
   errorcount, gtlength, ocrlength = levenshtein_multi_unicode(groundtruth, ocr)
   rate = float(errorcount) / gtlength
   print(("Errorcount:          %d" % errorcount))
   print(("Characters in GT:    %d" % gtlength))
   print(("Characters in OCR:   %d" % ocrlength))
   print(("Error Rate:          %.2f %%" % (rate * 100)))
   #print "=%5d %5d %5d %3.2f" % (errorcount, gtlength, ocrlength, rate*100)
   
   return rate



gamera/toolkits/greekocr/unicode_teubner.py
#encoding: utf-8
# -*- mode: python; indent-tabs-mode: nil; tab-width: 3 -*-
# vim: set tabstop=3 shiftwidth=3 expandtab:

# Copyright (C) 2010-2011 Christian Brandt
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import sys

charactermap = {
   "GREEK CAPITAL LETTER ALPHA": "A","GREEK CAPITAL LETTER BETA": "B",
   "GREEK CAPITAL LETTER GAMMA": "G","GREEK CAPITAL LETTER DELTA": "D",
   "GREEK CAPITAL LETTER EPSILON": "E","GREEK CAPITAL LETTER ZETA": "Z",
   "GREEK CAPITAL LETTER ETA": "H","GREEK CAPITAL LETTER THETA": "J",
   "GREEK CAPITAL LETTER IOTA": "I","GREEK CAPITAL LETTER KAPPA": "K",
   "GREEK CAPITAL LETTER LAMDA": "L","GREEK CAPITAL LETTER MU": "M",
   "GREEK CAPITAL LETTER NU": "N","GREEK CAPITAL LETTER XI": "X",
   "GREEK CAPITAL LETTER OMICRON": "O","GREEK CAPITAL LETTER PI": "P",
   "GREEK CAPITAL LETTER RHO": "R","GREEK CAPITAL LETTER SIGMA": "C",
   "GREEK CAPITAL LETTER TAU": "T","GREEK CAPITAL LETTER UPSILON": "U",
   "GREEK CAPITAL LETTER PHI": "F","GREEK CAPITAL LETTER CHI": "Q",
   "GREEK CAPITAL LETTER PSI": "Y","GREEK CAPITAL LETTER OMEGA": "W",
   "GREEK SMALL LETTER ALPHA": "a","GREEK SMALL LETTER BETA": "b",
   "GREEK SMALL LETTER GAMMA": "g","GREEK SMALL LETTER DELTA": "d",
   "GREEK SMALL LETTER EPSILON": "e","GREEK SMALL LETTER ZETA": "z",
   "GREEK SMALL LETTER ETA": "h","GREEK SMALL LETTER THETA": "j","GREEK THETA SYMBOL": "j",
   "GREEK SMALL LETTER IOTA": "i","GREEK SMALL LETTER KAPPA": "k",
   "GREEK SMALL LETTER LAMDA": "l","GREEK SMALL LETTER MU": "m",
   "GREEK SMALL LETTER NU": "n","GREEK SMALL LETTER XI": "x",
   "GREEK SMALL LETTER OMICRON": "o","GREEK SMALL LETTER PI": "p",
   "GREEK SMALL LETTER RHO": "r","GREEK SMALL LETTER FINAL SIGMA": "c",
   "GREEK SMALL LETTER SIGMA": "s","GREEK SMALL LETTER TAU": "t",
   "GREEK SMALL LETTER UPSILON": "u","GREEK SMALL LETTER PHI": "f",
   "GREEK SMALL LETTER CHI": "q","GREEK SMALL LETTER PSI": "y",
   "GREEK SMALL LETTER OMEGA": "w", "SPACE": " ", 
   "FULL STOP": ".", "COMMA": ",", "HYPHEN-MINUS": "-"
}

accentmap = [
   ['\\`%c', ['combining.grave.accent']], 
   ['\\\'%c', ['combining.acute.accent']], 
   ['\\~%c', ['combining.greek.perispomeni']], 
   ['\\"%c', ['combining.diaresis']], 
   ['\\u{%c}', ['combining.breve']], 
   ['\\U{%c%c}', ['combining.double.breve']], 
   ['\\=%c', ['combining.overline']], 
   ['\\r{%c}', ['combining.comma.above']], 
   ['\\s{%c}', ['combining.reversed.comma.above']], 
   ['\\Ad{%c}', ['combining.acute.accent', 'combining.diaresis']], 
   ['\\Gd{%c}', ['combining.diaresis', 'combining.grave.accent']], 
   ['\\Cd{%c}', ['combining.diaresis', 'combining.greek.perispomeni']], 
   ['\\Ar{%c}', ['combining.acute.accent', 'combining.reversed.comma.above']], 
   ['\\Gr{%c}', ['combining.grave.accent', 'combining.reversed.comma.above']], 
   ['\\Cr{%c}', ['combining.greek.perispomeni', 'combining.reversed.comma.above']], 
   ['\\As{%c}', ['combining.acute.accent', 'combining.comma.above']],
   ['\\Gs{%c}', ['combining.comma.above', 'combining.grave.accent']], 
   ['\\Cs{%c}', ['combining.comma.above', 'combining.greek.perispomeni']], 
   ['\\c{%c}', ['combining.inverted.breve.below']],
   ['\\ut{%cw}', ['combining.double.breve.below']],
   ['\\Ab{%c}', ['combining.acute.accent', 'combining.breve']],
   ['\\Gb{%c}', ['combining.breve', 'combining.grave.accent']],
   ['\\Arb{%c}', ['combining.acute.accent', 'combining.breve', 'combining.reversed.comma.above']],
   ['\\Grb{%c}', ['combining.breve', 'combining.grave.accent', 'combining.reversed.comma.above']],
   ['\\Asb{%c}', ['combining.acute.accent', 'combining.breve', 'combining.comma.above']],
   ['\\Gsb{%c}', ['combining.breve', 'combining.comma.above', 'combining.grave.accent']],
   ['\\Am{%c}', ['combining.acute.accent', 'combining.overline']],
   ['\\Gm{%c}', ['combining.grave.accent', 'combining.overline']],
   ['\\Cm{%c}', ['combining.greek.perispomeni', 'combining.overline']],
   ['\\Arm{%c}', ['combining.acute.accent', 'combining.overline', 'combining.reversed.comma.above']],
   ['\\Grm{%c}', ['combining.grave.accent', 'combining.overline', 'combining.reversed.comma.above']],
   ['\\Crm{%c}', ['combining.greek.perispomeni', 'combining.overline', 'combining.reversed.comma.above']],
   ['\\Asm{%c}', ['combining.acute.accent', 'combining.comma.above', 'combining.overline']],
   ['\\Gsm{%c}', ['combining.comma.above', 'combining.grave.accent', 'combining.overline']],
   ['\\Csm{%c}', ['combining.comma.above', 'combining.greek.perispomeni', 'combining.overline']],
   ['\\Sm{%c}', ['combining.comma.above', 'combining.overline']],
   ['\\Rm{%c}', ['combining.overline', 'combining.reversed.comma.above']],
   ['\\iS{%c}', ['combining.greek.ypogegrammeni']],
   ['\\d{%c}', ['combining.dot.below']],
   ['\\bd{%c}', ['combining.breve', 'combining.diaresis']],
   ['\\ring{%c}', ['combining.ring.below']]
]

accentmap.sort(key=lambda s: s[1])



def unicode_to_teubner(unicode_text):
   """Returns the given unicode string to a LaTeX document body using
the Teubner style for representing Greek characters and accents. Signature:

  ``unicode_to_teubner (unicode_text)``

The returned LaTeX code does not contain the LaTeX header. To create
a complete LaTeX document, you can use the following code:

.. code:: Python

   # LaTeX header
   print "\documentclass[10pt]{article}"
   print "\\usepackage[polutonikogreek]{babel}"
   print "\\usepackage[or]{teubner}"
   print "\\\\begin{document}"
   print "\selectlanguage{greek}"

   # document body
   print unicode_to_teubner(unicode_string)

   # LaTex footer
   print "\end{document}"

"""
   import unicodedata

   output = ""
   combinewith = []
   maincharacter = None
   i = 0
   while i < len(unicode_text):
      character = unicode_text[i]
      try:
         name_unicode = unicodedata.name(character)
      except:
         if character == "\n":
            output += " "

      name = name_unicode.lower()
      name = name.replace(" ", ".")
      
      
      if name.find("combining") == -1: #non-combining character
         if maincharacter != None and len(combinewith) > 0:
            #do lookup
            combinewith.sort()
            for format,combination in accentmap:
               if combination == combinewith:
                  try:
                     output += format % charactermap[maincharacter]
                  except KeyError:
                     sys.stderr.write("Teubner: Unknown character '%s'\n" % maincharacter)
                  break

            maincharacter = None
            combinewith = []

         elif maincharacter != None:
            try:
               output += charactermap[maincharacter]
               maincharacter = None
            except KeyError:
               sys.stderr.write("Teubner: Unknown character '%s'\n" % maincharacter)

         maincharacter = name_unicode
            
      else: #combining character
         if maincharacter != None:
            combinewith.append(name)
         else:
            output += "e"

      i += 1
   return output



if __name__ == "__main__":
   import unicodedata
   teststr = "ἔθαψε, ὡς οἰκὸς ἦν"


   print((unicode_to_teubner(unicodedata.normalize("NFD", teststr))))

   for a in accentmap:
      sort = sorted(a[1])
#unicodedata.normalize(u"aäb", "NFD"))


gamera/toolkits/greekocr/singlediacritics.py
# -*- mode: python; indent-tabs-mode: nil; tab-width: 3 -*-
# vim: set tabstop=3 shiftwidth=3 expandtab:

# Copyright (C) 2010-2011 Christian Brandt
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


from gamera.core import *    
from gamera.plugins.pagesegmentation import textline_reading_order
from gamera.toolkits.ocr.ocr_toolkit import *
from gamera.toolkits.ocr.classes import Textline,Page,ClassifyCCs
import gamera.kdtree as kdtree
import unicodedata
import sys

class SinglePage(Page):      
   def lines_to_chars(self):
      subccs = self.img.sub_cc_analysis(self.ccs_lines)
      
      for i,segment in enumerate(self.ccs_lines):
         self.textlines.append(SingleTextline(segment, subccs[1][i]))



class Character(object):
   def __init__(self, glyph):
      self.maincharacter = glyph
      self.unicodename = glyph.get_main_id()
      self.unicodename = self.unicodename.replace(".", " ").upper()
      #print self.unicodename
      #self.unicodename = 
      self.combinedwith = []
      #print self.maincharacter
      
   def addCombiningDiacritics(self, diacrit):
      self.combinedwith.append(diacrit)
      pass
   
   def toUnicodeString(self):
      try:
         str = ""
         mainids = self.maincharacter.get_main_id().split(".and.")
         for char in mainids:
            if char == "skip" or char == "unclassified":
               continue
            str = str + "%c" % return_char(char)
               
         #str = u"" + return_char(self.unicodename)
         for char in self.combinedwith:
            #char = char.get_main_id().replace(".", " ").upper()
            mainids = char.get_main_id().split(".and.")
            #print mainids
            for char in mainids:
               if char == "skip":
                  continue
               #print "added %s to output" % char
               str = str + "%c" % return_char(char)
      
         return unicodedata.normalize('NFD', str)
      except:
         #print self.unicodename
         return "E"
         
      
      
class SingleTextline(Textline):
   def sort_glyphs(self):
      self.glyphs.sort(lambda x,y: cmp(x.ul_x, y.ul_x))
      
      #begin calculating threshold for word-spacing
      glyphs = []
      for g in self.glyphs:
         if self.is_combining_glyph(g):
            continue
         
         glyphs.append(g)
         
      spacelist = []
      total_space = 0
      for i in range(len(glyphs) - 1):
         spacelist.append(glyphs[i + 1].ul_x - glyphs[i].lr_x)
      if(len(spacelist) > 0):
         threshold = median(spacelist)
         threshold = threshold * 2.0
      else:
         threshold  = 0
      #end calculatin threshold for word-spacing
      
      self.words = chars_make_words(self.glyphs, threshold)



   def is_combining_glyph(self, glyph):
      ret =  glyph.get_main_id().find("combining") != -1
      return ret
   


   def to_string(self):
      k = 3
      max_k = 10
      output = ""
      for word in self.words:
         med_center = median([g.center.y for g in word])
         glyphs_combining = []
         characters = []
         nodes_normal = []
         skipids = ["manual.xi.upper", "manual.xi.lower", "manual.theta.outer", "_split.splitx", "skip"]
         for glyph in word:
            mainid = glyph.get_main_id()
            
            if skipids.count(mainid) > 0:
               continue
            elif mainid == "manual.xi.middle":
               glyph.classify_automatic("greek.capital.letter.xi")
            elif mainid == "manual.theta.inner":
               glyph.classify_automatic("greek.capital.letter.theta")
            elif mainid == "comma" or mainid == "combining.comma.above":
               #print "%s - center_y: %d - med_center: %d" % (mainid, glyph.center.y, med_center)
               if glyph.center.y > self.bbox.center.y:
                  glyph.classify_automatic("comma")
               else:
                  glyph.classify_automatic("combining.comma.above")
            
            elif mainid.find("manual") != -1 or mainid.find("split") != -1:
               continue
            
            if self.is_combining_glyph(glyph):
               glyphs_combining.append(glyph)
            else:
               c = Character(glyph)
               characters.append(c)
               #print c
               nodes_normal.append(kdtree.KdNode((glyph.center.x, glyph.center.y), c))
         
         if (nodes_normal == None or len(nodes_normal) == 0):
            continue
            
         tree = kdtree.KdTree(nodes_normal)
         
         for g in glyphs_combining:
            fast = True
            if fast:
               knn = tree.k_nearest_neighbors((g.center.x, g.center.y), k)
               knn[0].data.addCombiningDiacritics(g)
            else:
               found = False
               while (not found) and k < max_k:
                  knn = tree.k_nearest_neighbors((g.center.x, g.center.y), k)
                  
                  for nn in knn:
                     if (nn.data.maincharacter.get_main_id().split(".").count("greek") > 0) and not found:
                        nn.data.addCombiningDiacritics(g)
                        found = True
                        break
               
                  k = k + 2      
                  
               
         for c in characters:
            output = output + c.toUnicodeString()
            
         output = output + " "



      return output



gamera/toolkits/greekocr/greekocr.py
# -*- mode: python; indent-tabs-mode: nil; tab-width: 3 -*-
# vim: set tabstop=3 shiftwidth=3 expandtab:

# Copyright (C) 2010-2011 Christian Brandt
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


from gamera.core import * 
init_gamera()    
from gamera import knn   
from gamera.plugins import pagesegmentation
from gamera.plugins.pagesegmentation import textline_reading_order
from gamera.classify import ShapedGroupingFunction
from gamera.plugins.image_utilities import union_images
from gamera.toolkits.ocr.ocr_toolkit import *
from gamera.toolkits.ocr.classes import Textline,Page,ClassifyCCs
from gamera.gamera_xml import glyphs_to_xml
from gamera.knn_editing import edit_cnn

from gamera.toolkits.greekocr.singlediacritics import *
from gamera.toolkits.greekocr.wholisticdiacritics import *
import unicodedata
import codecs



def clean_classifier(cknn):
   glyphs = cknn.get_glyphs()
   print(("old %d" % len(glyphs)))
   sorted_glyphs = sorted(glyphs, key=lambda g: g.to_rle())
   new_glyphs = []
   
   last_rle = sorted_glyphs[0].to_rle()
   new_glyphs.append(sorted_glyphs[0])
   for i in range(1, len(sorted_glyphs) -1):
         if last_rle != sorted_glyphs[i].to_rle():
            new_glyphs.append(sorted_glyphs[i])
         else:
            print((sorted_glyphs[i].get_main_id()))
   
   print(("new after removing duplicates: %d" % len(new_glyphs)))
   cknn.set_glyphs(new_glyphs)
   cknn = edit_cnn(cknn)
   
   print(("new after cnn: %d" % len(cknn.get_glyphs())))
   return cknn



class GreekOCR(object):
   """Provides the functionality for GreekOCR. The following parameters
control the recognition process:

  **cknn**
    The kNNInteractive classifier.

  **mode**
    The mode for dealing with accents.
    Can be ``wholistic`` or ``separatistic``.
"""
   def __init__(self, mode="wholistic"):
      """Signature:

   ``init (mode="wholistic")``

where *mode* can be "wholistic" or "separatistic".
"""
      self.optimizeknn = False
      self.debug = False
      self.cknn = knn.kNNInteractive([], ["aspect_ratio", "volume64regions", "moments", "nholes_extended"], 0)
      self.autogroup = None
      self.output = ""
      self.mode = mode
      


   def load_trainingdata(self, trainfile):
      """Loads the training data. Signature:

   ``load_trainingdata (trainfile)``

where *trainfile* is an Gamera XML file containing training data.
Make sure that the training file matches the *mode* (wholistic
or separatistic).
"""
      self.cknn.from_xml_filename(trainfile)
      if self.optimizeknn:
         self.cknn = clean_classifier(self.cknn)
     


   def segment_page(self):
      if(self.mode == "separatistic"):
         self.page = SinglePage(self.img)
      else:
         self.page = WholisticPage(self.img)

      if self.debug:
         print("start page segmentation...")
         t = time.time()

      self.page.segment()

      if self.debug:
         t = time.time() - t
         print(("\t segmentation done [",t,"sec]"))



   def get_page_glyphs(self, image):
      """Returns a list of segmented CCs using the selected segmentation 
approach on the given image. This list can be used for creating training data.
Signature:

   ``get_page_glyphs (image)``

where *image* is a Gamera image. 
"""
      if image.data.pixel_type != ONEBIT:
         image = image.to_onebit()
      self.img = image

      self.segment_page()
      glyphs = [] 
      for line in self.page.textlines:
         for g in line.glyphs:
            glyphs.append(g)

      return glyphs

      
   def save_debug_images(self):
      """Saves the following images to the current working directory:

  **debug_lines.png**
    Has a frame drawn around each detected line.

  **debug_chars.png**
    Has a frame drawn around each detected character.

  **debug_words.png**
    Has a frame drawn around each detected word.
"""
      rgbfilename = "debug_lines.png"
      rgb = self.page.show_lines()
      rgb.save_PNG(rgbfilename)
      print(("file '%s' written" % rgbfilename))
      rgbfilename = "debug_chars.png"
      rgb = self.page.show_glyphs()
      rgb.save_PNG(rgbfilename)
      print(("file '%s' written" % rgbfilename))
      rgbfilename = "debug_words.png"
      rgb = self.page.show_words()
      rgb.save_PNG(rgbfilename)
      print(("file '%s' written" % rgbfilename))



   def classify_text(self):
      self.output = ""
      for line in self.page.textlines:
         line.glyphs = \
            self.cknn.classify_and_update_list_automatic(line.glyphs)
         line.sort_glyphs()
         self.output = self.output + line.to_string() + "\n"

      self.output = self._normalize(self.output) 



   def get_text(self):
      return self.output



   def process_image(self, image):
      """Recognizes the given image and returns the recognized text
as Unicode string. Signature:

   ``process_image (image)``

where *image* is a Gamera image. The recognized text is additionally stored
in the ``GreekOCR`` property *output*, which can subsequently be written to
a file with save_text_unicode_ or save_text_teubner_.

Make sure that you have called load_trainingdata_ before!
"""
      if image.data.pixel_type != ONEBIT:
         image = image.to_onebit()
      self.img = image
      if self.debug:
         print("Doing page Segmentation")
      self.segment_page()
      if self.debug:
         print("Classifying Text")
      self.classify_text()
      if self.debug:
         print("Returning Output")
      return self.get_text()
      


   def save_text_xetex(self, filename):
      data = \
'''\documentclass[11pt]{article}
\\usepackage{xltxtra}
\setmainfont[Mapping=tex-text]{GFS Porson}
\\begin{document}
%s
\end{document}''' % self.output.replace("\n", "\n\n")

      f = codecs.open(filename, "w", encoding='utf-8')
      f.write(data)
      f.close()
      


   def save_text_unicode(self, filename):
      """Stores the recognized text to the given *filename* as Unicode string.
Signature

   ``save_text_unicode(filename)``

Make sure that you have called process_image_ before!
"""
      f = codecs.open(filename, "w", encoding='utf-8')
      f.write(self.output)
      f.close()



   def save_text_teubner(self, filename):
      """Stores the recognized text to the given *filename* as a LaTeX
document utilizing the Teubner style for representing Greek characters and
accents.
Signature

   ``save_text_teubner(filename)``

Make sure that you have called process_image_ before!
"""
      from .unicode_teubner import unicode_to_teubner

      data = '''
\documentclass[10pt]{article}
\\usepackage[polutonikogreek]{babel}
\\usepackage[or]{teubner}
\\begin{document}
\selectlanguage{greek}
%s
\end{document}
''' % unicode_to_teubner(self.output).replace("\n", "\n\n")

      f = codecs.open(filename, "w", encoding='utf-8')
      f.write(data)
      f.close()



   
   def _normalize(self,str):
      str = unicodedata.normalize("NFD", str)
      output = ""
      combined = []
      for i in str:
         is_combining = True
         try:
            is_combining = unicodedata.combining(i) > 0 or unicodedata.name(i).find("ACCENT") >= 0
         except:
            is_combining = False
         if not is_combining:
            for j in sorted(combined):
               output += j
            combined = []
            output += i
         else:
            combined.append(i)
   
      if len(combined) > 0:
         for j in sorted(combined):
            output += j
      return unicodedata.normalize("NFD", output)

